keras_rs/src/layers/embedding/embed_reduce.py

```diff
@@ -0,0 +1,188 @@
+        self.combiner = combiner
+
+    def call(
+        self,
+        inputs: types.Tensor,
+        weights: Optional[types.Tensor] = None,
+    ) -> types.Tensor:
+        """Apply embedding and reduction.
+
+        Args:
+            inputs: 1D tensor to embed or 2D tensor to embed and reduce.
+            weights: Optional tensor of weights to apply before reduction, which
+               is 2D and the same shape as `inputs`.
+
+        Returns:
+            A dense 2D tensor of shape `(batch_size, output_dim)`.
+        """
+        x = super().call(inputs)
+        unreduced_rank = len(x.shape)
+
+        # Check that weights has a compatible shape.
+        if weights is not None:
+            weights_rank = len(weights.shape)
+            if weights_rank > unreduced_rank or not check_shapes_compatible(
+                x.shape[0:weights_rank], weights.shape
+            ):
+                raise ValueError(
+                    f"The shape of `weights`: {weights.shape} is not compatible"
+                    f" with the shape of `inputs` after embedding: {x.shape}."
+                )
+
+        dtype = (
+            x.dtype
+            if weights is None
+            else keras.backend.result_type(x.dtype, weights.dtype)
+        )
+
+        # When the inputs are 1D (unreduced_rank == 2), no reduction is applied,
+        # so we want to short circuit the reduction logic.
+        # With nothing to reduce, the "mean" and "sqrtn" combiners don't modify
+        # the result, but we still use the weights as a mask for the result for
+        # ragged and sparse.
+        # With the "sum" combiner, even without reduction, we still need to
+        # multiply by the weight to produce the correct result.
+        if weights is None or (unreduced_rank <= 2 and self.combiner != "sum"):
+            # Discard the weights if there were some and create a mask for
+            # ragged and sparse tensors to mask the result correctly (sparse
+            # only) and the apply the reduction correctly (ragged and sparse).
+            weights = ops.ones_like(inputs, dtype=dtype)
+        else:
+            weights = ops.cast(weights, dtype)
+
+        # When looking up using sparse indices, the result is dense but contains
+        # values that should be ignored as all missing values use index 0. We
+        # use `weights` as a mask, but it needs to be densified as
+        # `expand_dims` and broadcasting a sparse tensor does not produce the
+        # expected result.
+        weights = ops.convert_to_tensor(weights, sparse=False)
+
+        # Make weights and the unreduced embeddings have the same rank.
+        weights_rank = len(weights.shape)
+        if weights_rank < unreduced_rank:
+            weights = ops.expand_dims(
+                weights, axis=tuple(range(weights_rank, unreduced_rank))
+            )
+
+        # Note that `x` and `weights` are:
+        # - ragged if `inputs` was ragged and `weights` was ragged or None
```

Comment:
How can a tensor be ragged since JAX and Torch do not have ragged support? Confused

Reply:
Oh, this if for TF only for now. This comment is specifically to explain that it's usually dense.

Note that this code was carefully crafted to handle dense, ragged and sparse without branching and without even knowing about the concept of ragged / sparse. (Part of it is we're missing APIs for that)

Reply:
Ah, okay. I should have checked the unit tests